File: tests/r_parity/28_frontier.py
"""StatsPAI cross-section stochastic frontier parity (Python side).
Module 28.

Generates a deterministic Cobb-Douglas frontier DGP with half-normal
inefficiency. The companion 28_frontier.R uses sfaR::sfacross with
the same distribution. Tolerance: rel < 1e-6 on production-frontier
point estimates after the tightened frontier optimizer; sigma SE rows
remain on a separate convention guard.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "28_frontier"


def make_data(n: int = 500, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    lnk = rng.normal(0, 1, n)
    lnl = rng.normal(0, 1, n)
    v = rng.normal(0, 0.3, n)
    u = np.abs(rng.normal(0, 0.5, n))
    lny = 2.0 + 0.6 * lnk + 0.4 * lnl + v - u
    return pd.DataFrame({"lny": lny, "lnk": lnk, "lnl": lnl})


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    fit = sp.frontier(data=df, y="lny", x=["lnk", "lnl"],
                      dist="half-normal", seed=PARITY_SEED)

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "beta_intercept",
                     estimate=float(fit.params["_cons"]),
                     se=float(fit.std_errors["_cons"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "beta_lnk",
                     estimate=float(fit.params["lnk"]),
                     se=float(fit.std_errors["lnk"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "beta_lnl",
                     estimate=float(fit.params["lnl"]),
                     se=float(fit.std_errors["lnl"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "sigma_u",
                     estimate=float(fit.model_info["sigma_u_mean"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "sigma_v",
                     estimate=float(fit.model_info["sigma_v_mean"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "lambda",
                     estimate=float(fit.model_info["lambda"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "mean_efficiency_jlms",
                     estimate=float(fit.model_info["mean_efficiency_jlms"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "mean_efficiency_bc",
                     estimate=float(fit.model_info["mean_efficiency_bc"]),
                     n=int(len(df))),
    ]

    write_results(MODULE, "py", rows,
                  extra={"distribution": "half-normal",
                         "cost": False,
                         "method": fit.model_info["method"],
                         "efficiency_note": (
                             "R sfaR::efficiencies(fit)[,'teJLMS'] and "
                             "StatsPAI mean_efficiency_jlms are JLMS "
                             "exp(-E[u|eps]); Stata frontier predict, te "
                             "reports the Battese-Coelli conditional "
                             "E[exp(-u)|eps] scale, exposed as "
                             "mean_efficiency_bc."
                         )})


if __name__ == "__main__":
    main()

File: tests/r_parity/28_frontier.R
# StatsPAI cross-section frontier parity (R side) -- Module 28.
#
# Reads data/28_frontier.csv and runs sfaR::sfacross with half-
# normal inefficiency. Tolerance: rel < 1e-6 on production-frontier
# point estimates; sigma SE rows remain on a separate convention guard.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(sfaR)
})

MODULE <- "28_frontier"

df <- read_csv_strict(MODULE)

set.seed(PARITY_SEED)
fit <- sfaR::sfacross(
  formula = lny ~ lnk + lnl,
  data    = df,
  udist   = "hnormal",
  S       = 1L              # production frontier (S = -1 for cost)
)

# sfaR returns coefficients in `mlParam`; the frontier coefs come
# first, then the variance parameters in log-form.
co <- coef(fit)
# Standard errors:
se <- sqrt(diag(vcov(fit)))

# Variance components: sfaR exposes sigmaSq_u, sigmaSq_v.
em <- summary(fit)$mlParamMatrix
# Extract sigma_u, sigma_v from the log-parametrisation.
ln_sigma_u_sq <- co["Wu_(Intercept)"]
ln_sigma_v_sq <- co["Vu_(Intercept)"]
sigma_u <- exp(0.5 * ln_sigma_u_sq)
sigma_v <- exp(0.5 * ln_sigma_v_sq)
lambda  <- sigma_u / sigma_v

# Mean technical efficiency (Battese-Coelli).
eff <- sfaR::efficiencies(fit)
mean_eff_jlms <- mean(eff[, "teJLMS"])  # Jondrow-Lovell-Materov-Schmidt
mean_eff_bc <- if ("teBC" %in% colnames(eff)) mean(eff[, "teBC"]) else NA_real_

rows <- list(
  parity_row(MODULE, "beta_intercept",
             estimate = unname(co["(Intercept)"]),
             se = unname(se["(Intercept)"]), n = nrow(df)),
  parity_row(MODULE, "beta_lnk",
             estimate = unname(co["lnk"]),
             se = unname(se["lnk"]), n = nrow(df)),
  parity_row(MODULE, "beta_lnl",
             estimate = unname(co["lnl"]),
             se = unname(se["lnl"]), n = nrow(df)),
  parity_row(MODULE, "sigma_u", estimate = sigma_u, n = nrow(df)),
  parity_row(MODULE, "sigma_v", estimate = sigma_v, n = nrow(df)),
  parity_row(MODULE, "lambda", estimate = lambda, n = nrow(df)),
  parity_row(MODULE, "mean_efficiency_jlms",
             estimate = mean_eff_jlms, n = nrow(df))
)
if (is.finite(mean_eff_bc)) {
  rows[[length(rows) + 1L]] <- parity_row(
    MODULE, "mean_efficiency_bc",
    estimate = mean_eff_bc, n = nrow(df))
}

write_results(MODULE, rows,
              extra = list(distribution = "half-normal",
                           S = 1, package = "sfaR::sfacross",
                           efficiency_note = paste(
                             "mean_efficiency_jlms is teJLMS;",
                             "mean_efficiency_bc is teBC when exposed by sfaR.")))
